nemo_aligner/utils/trt_llm.py

```diff
@@ -0,0 +1,230 @@
+        seed=None,
+        unload_engine_train=False,
+        reshard_model=False,
+        trt_model_dir="/tmp/trt_llm_model",
+    ):
+        assert (
+            tokenizer.pad_id != tokenizer.eos_id
+        ), "We require tokenizers to have a different pad_id than eos_id when using TRT-LLM. This is to make sure all code goes into the same path and include the eos_id when the response lengths are computed"
+
+        if use_greedy and sample_top_k != 1:
+            logging.warning(f"'use_greedy=True' overrides {sample_top_k=} to 1")
+            sample_top_k = 1
+
+        self.model_cfg = model_cfg
+        self.tokenizer = tokenizer
+        self.max_generation_length = max_generation_length
+        self.max_input_len = max_input_len
+        self.max_input_tokens = max_input_tokens
+        self.generation_batch_size = generation_batch_size
+        self.unload_engine_train = unload_engine_train
+        self.trt_model_type = trt_model_type
+        self.reshard_model = reshard_model
+        self.trt_llm_exporter = TensorRTLLM(trt_model_dir, load_model=False)
+        self._trtllm_model_compiled = False
+
+        rng_generator = torch.Generator(device="cpu")
+        seed = secrets.randbits(32) if seed is None else seed
+        rng_generator.manual_seed(seed)
+        self.rng_generator = rng_generator
+
+        self.pad_id = tokenizer.pad_id
+        end_id = tokenizer.eos_id
+        end_strings = list(end_strings)
+
+        # TRT-LLM uses different logic to compute the response length depending on if we specify stop_list or end_id
+        # we want to use the same logic (which is to include the stop token in response length) so we put end_id into the stop_list
+        # manually
+        if len(end_strings) == 0:
+            ids, offsets = [end_id], [1]
+        else:
+            assert all(
+                "," not in end_string for end_string in end_strings
+            ), "detected `,` in a specified end_string. This is will cause an error when converting it into ids for TRT-LLM"
+
+            end_strings = [[",".join(end_strings)]]
+            # use an arbitary ref string to obtain end_string ids
+            stop_list = tensorrt_llm_run.to_word_list_format(
+                end_strings, build_tokenizer(self.tokenizer), ref_str="green tea icecream"
+            )
+            ids, offsets = stop_list[0].tolist()
+            # add the end_id if it doesn't exist
+            if end_id not in ids:
+                ids = append_and_repad_list(ids, end_id, pad_id=0)
+                offsets = append_and_repad_list(offsets, max(offsets) + 1, pad_id=-1)
+
+        assert max(offsets) == len(ids), "offset and stop token length are mismatched"
+        stop_list = torch.as_tensor([ids, offsets], dtype=torch.int32, device=torch.cuda.current_device()).repeat(
+            self.generation_batch_size, 1, 1
+        )
+
+        self.sampling_config = tensorrt_llm.runtime.SamplingConfig(
+            end_id=self.pad_id,
```

Comment:
```suggestion
            # We use `pad_id` as end token instead of `eos_id` to actually "disable" this end token
            # mechanism. Otherwise, the response length returned by TRT-LLM would *not* count `eos_id`
            # when the model would end its generation with this token. Instead, `stop_words_list` is
            # augmented with `eos_id` which ensures it is counted in the response length (see above).
            end_id=self.pad_id,
```